app/src/org/commcare/tasks/ProcessAndSendTask.java

```diff
@@ -174,60 +177,68 @@ private boolean checkFormRecordStatus(FormRecord[] records)
 
             //If the form is complete, but unprocessed, process it.
             if (FormRecord.STATUS_COMPLETE.equals(record.getStatus())) {
+                SQLiteDatabase db =
+                        CommCareApplication.instance().getUserDbHandle();
+                db.beginTransaction();
                 try {
                     records[i] = processor.process(record);
+                    db.setTransactionSuccessful();
                 } catch (InvalidStructureException | XmlPullParserException |
                         UnfullfilledRequirementsException e) {
-                    handleExceptionFromFormProcessing(record, e);
+                    db.endTransaction();

```

Comment:
One other option for this pattern would be 

```
try{
  toRethrow = null
  beginTransaction
  try{
    stuff
    transactionSuccesful
  } catch(Exception e) {
    toRethrow = e
  } finally {
    endtransaction
  }
  if(rethrow != null) { throw rethrow }

} catch(actualerrorhandlinghere) {
}

````

Reply:
@ctsims I liked how this looked in theory and just tried to implement it, but there's actually a compilation problem with it. In order to do this, we would have to change the method signature to include `throws Exception` since the compiler now interprets this method as being capable of throwing any type of Exception at all. I'm not convinced that that's a good trade-off, so I'm going to leave this the way I had it for now.

Reply:
Hey, 

I was actually making that way more complex than it needed to be.

All you actually need to do here is
```
try {
  startTransaction()
  try {
    doCode();
    transactionSuccess();
  } finally {
    endTransaction();
  } 
} catch(TypedException e) {
  handle stuff
}
```

The interior finally() block will execute as soon as java leaves the scope of the interior try{} block before the exterior block takes the  control flow